be/src/column/map_column.cpp

```diff
@@ -422,34 +423,75 @@ int MapColumn::equals(size_t left, const Column& rhs, size_t right, bool safe_eq
         return false;
     }
 
-    bool has_null = false;
+    // process the null key at last if exists, so non-nullable keys can exactly identify equal one or not.
+    // if any non-nullable key does not match, return false;
+    // else if all non-nullable key are matched (true or null), check the last nullable keys.
+    // if the last nullable key is not matched, return false; else if there is null result from all keys matching,
+    // return null, else return true.
+
+    bool has_null_eq = false;
+    uint32_t null_id = 0;
+    std::vector<uint32_t> index;
     for (uint32_t i = lhs_offset; i < lhs_end; ++i) {
-        bool found = false;
-        for (uint32_t j = rhs_offset; j < rhs_end; ++j) {
-            int res = _keys->equals(i, *(rhs_map._keys.get()), j, safe_eq);
-            if (res == EQUALS_FALSE) {
+        if (_keys->is_null(i)) {
+            null_id = i;
+            continue;
+        }
+        index.push_back(i);
+    }
+    if (index.size() < (lhs_end - lhs_offset)) {
+        index.push_back(null_id);
+    }
+    std::set<uint32_t> right_index;
+    for (uint32_t j = rhs_offset; j < rhs_end; ++j) {
+        right_index.insert(j);
+    }
+
+    for (auto i : index) {
+        bool real_eq = false;
+        bool null_eq = false;
+        uint32_t eq_id = 0;
+        for (unsigned int j : right_index) {
+            int key_res = _keys->equals(i, *(rhs_map._keys.get()), j, safe_eq);
+            if (key_res == EQUALS_FALSE) {
                 continue;
             }
-
-            has_null |= (res == EQUALS_NULL);
-            // So two keys is the same
-            res = _values->equals(i, *(rhs_map._values.get()), j, safe_eq);
-            if (res == EQUALS_FALSE) {
-                return EQUALS_FALSE;
+            // So two keys are the same or right key is null
+            int val_res = _values->equals(i, *(rhs_map._values.get()), j, safe_eq);
+
+            // case 1: key_res == EQUALS_TRUE
+            if (key_res == EQUALS_TRUE) {
```

Comment:
The `if (key_res == EQUALS_TRUE) {` is non-sense.

Reply:
not true, it has to diff the case where `key_res == EQUAL_NULL`